File: hms_tz/nhif/api/therapy_plan.py
import frappe

def before_insert(doc):
    total_sessions = 0
    for entry in doc.therapy_plan_details:
        if entry.no_of_sessions:
            total_sessions += entry.no_of_sessions
    doc.total_sessions =  total_sessions

def validate(doc, method):
    set_totals(doc)
    set_status(doc)

def set_status(doc):
    if (
        not doc.total_sessions_completed and 
        not doc.total_sessions
    ):
        doc.status = 'Not Serviced'
    
    if (
        doc.total_sessions and 
        not doc.total_sessions_completed
    ):
        doc.status = 'Not Started'

    else:
        if doc.total_sessions_completed < doc.total_sessions:
            doc.status = 'In Progress'

        elif (
            doc.total_sessions != 0 and 
            (doc.total_sessions_completed == doc.total_sessions)
        ):
            doc.status = 'Completed'

def set_totals(doc):
    total_sessions_completed = 0
    for entry in doc.therapy_plan_details:
        if entry.sessions_completed:
            total_sessions_completed += entry.sessions_completed

    doc.db_set('total_sessions_completed', total_sessions_completed)

File: hms_tz/nhif/api/therapy_plan.js
frappe.ui.form.on('Therapy Plan', {
	setup: function(frm) {
		frm.get_field('therapy_plan_details').grid.editable_fields = [
			{fieldname: 'therapy_type', columns: 6},
			{fieldname: 'no_of_sessions', columns: 2},
			{fieldname: 'sessions_completed', columns: 2}
		];
	},
	
	refresh: function(frm) {
		if (!frm.doc.__islocal && (frm.doc.status == 'Not Serviced')) {
			frm.clear_custom_buttons();
			// frm.remove_custom_button("Create")
		}
		if (frm.doc.patient) {
			frm.add_custom_button(__('Patient History'), () => {
				frappe.route_options = { 'patient': frm.doc.patient };
				frappe.set_route('tz-patient-history');
			});
		}
	},
});